core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2060,6 +2060,25 @@ private RexNode simplifyOrs(List<RexNode> terms, RexUnknownAs unknownAs) {
           }
         }
         break;
+      case IS_NOT_TRUE:
+        if (terms.contains(((RexCall) term).getOperands().get(0))) {
+          return trueLiteral;
+        }
+        break;
+      case NOT:
+        RexNode x = ((RexCall) term).getOperands().get(0);
+        if (terms.contains(x)) {
```

Comment:
I think to proceed in both cases `x` must be `isSafeExpression` to remove them

Reply:
@kgyrtkirk Thanks for your review! 
All operands in `OR` operator must be `boolean` type (base on the operand checker), `x` could be `1 / 0 <> 0` but could not be `1 / 0`. So I think wo don't need to care about if `x` safe or not. WDYT?

Reply:
you've replied to that comment here..that's not great - because I think these are 2 separate issues - but they are somewhat similar.

`1/0` is not a boolean - my mistake; I just wanted to illustrate the problem quickly

Don't expect that `1 / 0` will appear as is ... it will be more likely `x / y > 17` which is a boolean; can't be decided at compile time that it will never or always throw exceptions - but still holds the possiblity of an exception because if `y` is `0` it will emit an exception.

I think the current patch will  remove and simplify unknown udfs away - we don't know for sure that  omitting those calls remove sideeffects or not.

I think
```
someUdf() or someUdf() is not true  ==> true
```
is wrong because it could be a possibly unsafe method - you may not make such transformation; or at least call that method - but I would rather stay on the safe side and not do that either...

and also
```
someUdf() or x or !x ==> true (in case x is not nullable)
```
almost the same as above - however in this case the unsafe expr is outside

In the above cases: what if `someUdf` throws an exception on every invocation? should the query be executed - because we were able to deduce that its return value is irrelevant?


Could you please add some illustrative testcases for the above cases?

Reply:
Thanks for your explaination, I got your points. 
For the first case `x or not(x)`, I add a new limitation where `x` must be safeExpression.
But for the second case `a or x or not(x) or b`, I think we can simplify `x or not(x)` no matter whether `a` and `b` safe or not.
Thinking about a simple case: `true or unSafeUdf()`,  calcite will simplify this expression to `true`. And I have tested `select * from t where 1/0 or true` expression in mysql and sparksql, both of them will simpilfy this expression to `true`. And third, in most of programing language (like Java, C++, Python), they always simplify `true or someConditions` to `true` in if clause (A little different, it is Short-circuit OR actully, but we check all items).
So I think we don't need to care about outside items whether safe or not.